lib/ZcapClient.js

```diff
@@ -0,0 +1,228 @@
+   *
+   * Creates a new ZcapClient instance that can be used to perform
+   * Authorization Capability (ZCAP) requests against HTTP URLs.
+   *
+   * @param {object} options - The options to use.
+   * @param {string} options.baseUrl - The base URL for the client to use
+   *   when building invocation request URLs.
+   * @param {object} [options.defaultHeaders] - The optional default HTTP
+   *   headers to include in every invocation request.
+   * @param {object} [options.agent] - An optional HttpsAgent to use to
+   *   when performing HTTPS requests.
+   * @param {object} options.invocationSigner - An object with a
+   *   `.sign()` function and key `id` parameter that will be used for
+   *   signing the request.
+   * @param {object} options.delegationSigner - An object with a
+   *   `.sign()` function and key `id` parameter that will be used for
+   *   signing a capability delegation.
+   *
+   * @returns {ZcapClient} - The new ZcapClient instance.
+   */
+  constructor({
+    baseUrl, defaultHeaders = {}, agent, invocationSigner, delegationSigner
+  } = {}) {
+    this.baseUrl = baseUrl;
+    this.agent = agent;
+    this.defaultHeaders = {...DEFAULT_HEADERS, ...defaultHeaders};
+    this.invocationSigner = invocationSigner;
+    this.delegationSigner = delegationSigner;
+  }
+
+  /**
+   * Delegates an Authorization Capability to a specified delegatee.
+   *
+   * @param {object} options - The options to use.
+   * @param {string} [options.url] - The relative URL to invoke the
+   *   Authorization Capability against. Either `url` or `capability` must
+   *   be specified.
+   * @param {string} [options.capability] - The relative URL to invoke the
+   *   Authorization Capability against. Either `capability` or `url` must
+   *   be specified.
+   * @param {string} options.delegate - The URL identifying the entity to
+   *   delegate to.
+   * @param {string} [options.expires] - Optional expiration value for the
+   *   delegation. Default is 5 minutes into the future.
+   * @param {string|Array} [options.allowedActions] - Optional list of allowed
+   *   actions or string specifying allowed delegated action. Default: [] -
+   *   delegate all actions.
+   *
+   * @returns {object} - A promise that resolves to an HTTP response.
+   */
+  async delegate({
+    url, capability, delegate, expires, allowedActions = []
+  } = {}) {
+    let delegatedCapability;
+
+    // convert string value for allowedActions to array
+    allowedActions = (typeof allowedActions === 'string') ?
+      allowedActions = [allowedActions] : allowedActions;
+
+    // default expiration is 5 minutes in the future
+    const defaultExpires = new Date(Date.now() + 5 * 60 * 1000).toISOString();
+
+    if(url) {
+      // generate the root capability
+      delegatedCapability = {
+        '@context': SECURITY_CONTEXT_V2_URL,
+        id: `urn:zcap:${await generateId()}`,
+        parentCapability: 'urn:zcap:' + url,
```

Comment:
```suggestion
        parentCapability: `urn:zcap:${url}`,
```

Also, we need to decide if we want to do URL encoding here. We probably do, right?